### PR TITLE
[RW-7893][no risk]lock workspace test

### PR DESCRIPTION
**Pull request by chavanr2019**
TEST GOAL
 Verify if lock workspace feature works as expected.

- Test case 1: verify if workspace is locked
- Test case 2: verify if workspace is unlocked.

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    await aboutPage.openAnalysisPage({ waitPageChange: false });
+    // verify share button is disabled
+    expect(await aboutPage.getShareButton().isCursorNotAllowed()).toBe(true);
+    await aboutPage.verifyLockedWorkspaceActionOptions();
+    // verify the user is able to navigate to edit page to update the description
+    await aboutPage.getAboutEditIcon();
+    const workspaceEdit = new WorkspaceEditPage(page);
+    await workspaceEdit.waitForLoad();
+    // verify user is able to edit the locked workspace
+    workspaceEditedName = await workspaceEdit.fillOutRequiredDuplicationFields();
```
Comment (chavanr2019) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts:78]: @aweng98 Since I'm reusing this function `fillOutRequiredDuplicationFields()`, I suggest we update the name to editWorkspaceName.

**Review by chavanr2019 — COMMENTED** [review]
@jmthibault79 @aweng98 Please review, Thanks!

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 2]
```diff
@@ -65,8 +66,8 @@ describe('Workspace Admin lock-workspace', () => {
     const aboutLockedIcon = aboutPage.getAboutLockedWorkspaceIcon();
     expect(aboutLockedIcon).toBeTruthy();
     // verify DATA & ANALYSIS tabs are disabled
-    await aboutPage.openDataPage({ waitPageChange: false });
-    await aboutPage.openAnalysisPage({ waitPageChange: false });
+    // await aboutPage.openDataPage({ waitPageChange: false });
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts]: why this change?

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 3]
```diff
@@ -65,9 +65,9 @@ describe('Workspace Admin lock-workspace', () => {
     expect(aboutLockReason).toEqual(reasonText);
     const aboutLockedIcon = aboutPage.getAboutLockedWorkspaceIcon();
     expect(aboutLockedIcon).toBeTruthy();
-    // verify DATA & ANALYSIS tabs are disabled
-    // await aboutPage.openDataPage({ waitPageChange: false });
-    // await aboutPage.openAnalysisPage({ waitPageChange: false });
+    // verify DATA & ANALYSIS tabs are inactive for locked workspace
+    expect(await aboutPage.getTabsAttribute(page, Tabs.Analysis, 'aria-selected')).toEqual('false');
+    expect(await aboutPage.getTabsAttribute(page, Tabs.Data, 'aria-selected')).toEqual('false');
```
Comment (aweng98) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts:70]: `Tab` has `isSelected()` which you can use for this check.

File: e2e/tests/nightly/admin/lock-workspace.spec.ts (hunk 2 above)
Comment (chavanr2019) [reply e2e/tests/nightly/admin/lock-workspace.spec.ts]: Because of the changes that happened in this file: [workspace-base.ts](https://github.com/all-of-us/workbench/pull/6276/files#diff-f0d5c7b7e18c9b2a759b0324fb48ae1be874e4e02bcdd4dd58e8a9eb38deaff1)

File: e2e/tests/nightly/admin/lock-workspace.spec.ts (hunk 3 above)
Comment (chavanr2019) [reply e2e/tests/nightly/admin/lock-workspace.spec.ts:70]: ok! will try. Thanks!

**Review by chavanr2019 — COMMENTED** [review]
@jmthibault79 the test steps at [line 69-70](https://github.com/all-of-us/workbench/pull/6326/files#diff-a8ed241affb7745365cbc122d7e282e43bb2fd9348a294ef90a8c61bf25bd9bfR69-R70) have been updated since this file has been refactored: [workspace-base.ts](https://github.com/all-of-us/workbench/pull/6276/files#diff-f0d5c7b7e18c9b2a759b0324fb48ae1be874e4e02bcdd4dd58e8a9eb38deaff1)

File: e2e/app/page/workspace-about-page.ts [hunk 4]
```diff
@@ -73,4 +77,46 @@ export default class WorkspaceAboutPage extends WorkspaceBase {
...
+    await edit.click();
+    const editPage = new WorkspaceEditPage(page);
+    return await editPage.waitForLoad();
+  }
+
+  // verify the state of the inactive tabs
+  async getTabState(page: Page, tabName: Tabs): Promise<boolean> {
+    const tab = new Tab(page, tabName);
+    await tab.isSelected();
+    return false;
```
Comment (chavanr2019) [inline e2e/app/page/workspace-about-page.ts:120]: @aweng98 used `isSelected()` to verify the tab state.

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    // verify the banner includes the locking reason
+    const aboutLockReason = await aboutPage.extractReasonMessage();
+    expect(aboutLockReason).toEqual(reasonText);
+    // verify the lock icon is displaying for the locked workspace
+    const aboutLockedIcon = aboutPage.getAboutLockedWorkspaceIcon();
+    expect(aboutLockedIcon).toBeTruthy();
+
+    // verify DATA & ANALYSIS tabs are inactive for locked workspace
+    expect(await aboutPage.getTabState(page, Tabs.Analysis)).toBe(false);
+    expect(await aboutPage.getTabState(page, Tabs.Data)).toBe(false);
```
Comment (chavanr2019) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts:74]: @aweng98 refactored the tab state verification.

File: e2e/app/page/workspace-about-page.ts [hunk 6]
```diff
@@ -113,9 +113,11 @@ export default class WorkspaceAboutPage extends WorkspaceBase {
...
 
-  // get the tab attribute to verify the tabs state
-  async getTabsAttribute(page: Page, tabName: Tabs, attribute: string): Promise<string> {
-    const tab = new Tab(page, tabName);
-    return await tab.getAttribute(attribute);
+  // verify the state of the inactive tabs
+  async getTabState(page: Page, tabName: Tabs): Promise<boolean>{
+      const tab = new Tab(page, tabName);
+      await tab.isSelected();
+      return false;
```
Comment (jmthibault79) [inline e2e/app/page/workspace-about-page.ts]: this *always* returns false.  Not what we want.

Comment (chavanr2019) [reply e2e/app/page/workspace-about-page.ts]: ur right! I'll look into it. Thanks!

File: e2e/app/page/workspace-about-page.ts [hunk 7]
```diff
@@ -73,4 +77,46 @@ export default class WorkspaceAboutPage extends WorkspaceBase {
...
+    await edit.click();
+    const editPage = new WorkspaceEditPage(page);
+    return await editPage.waitForLoad();
+  }
+
+  // get the state of the  tabs
+  async getTabState(tabName: Tabs): Promise<boolean> {
+    const tab = new Tab(this.page, tabName);
+    const tabState = await tab.isSelected();
+    return tabState;
```
Comment (chavanr2019) [inline e2e/app/page/workspace-about-page.ts:120]: @aweng98 I've used `isSelected()` but it is returning undefined instead of boolean. what do you think is happening here, is it function not right or I'm missing something?

**Review by chavanr2019 — COMMENTED** [review]
@jmthibault79 I've updated the `getTabState`but it is returning me undefined instead of boolean, not sure what is happening. Please let me know what I'm missing. Thanks!

**Review by jmthibault79 — COMMENTED** [review]
Looks good except for the tab-selection problem ... I'll wait for you to work that out

File: e2e/app/page/workspace-about-page.ts [hunk 8]
```diff
@@ -73,4 +77,46 @@ export default class WorkspaceAboutPage extends WorkspaceBase {
...
+    const messageText = (await reasonMessage).slice(8);
+    return messageText;
+  }
+
+  // click the edit icon on about page to navigate to the edit page
+  async getAboutEditIcon(): Promise<WorkspaceEditPage> {
+    const xpath = '//div[contains(text(), "Primary purpose of project")]/div[@role="button"]';
+    const edit = new Button(this.page, xpath);
+    await edit.click();
+    const editPage = new WorkspaceEditPage(page);
```
Comment (aweng98) [inline e2e/app/page/workspace-about-page.ts:112]: Don't use global object `page`. It should be `this.page`. Can you double check all other usages?

File: e2e/tests/nightly/admin/lock-workspace.spec.ts (hunk 1 above)
Comment (aweng98) [reply e2e/tests/nightly/admin/lock-workspace.spec.ts:86]: No, `fillOutRequiredDuplicationFields()` is fine.

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    await workspaceCard.verifyWorkspaceCardMenuOptions();
+    await workspaceCard.clickWorkspaceName(true);
+    const dataPage = new WorkspaceDataPage(page);
+    await dataPage.waitForLoad();
+    // verify DATA & ANALYSIS tabs are aactive and accessible
+    const analysisPage = new WorkspaceAnalysisPage(page);
+    await openTab(page, Tabs.Analysis, analysisPage);
+    const aboutPage = new WorkspaceAboutPage(page);
+    await openTab(page, Tabs.About, aboutPage);
+    await aboutPage.waitForLoad();
```
Comment (aweng98) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts:115]: Check out the `openTab` function implementation. You will see that `await aboutPage.waitForLoad()` is not needed. Can you remove unneeded `waitForLoad` everywhere?

File: e2e/tests/nightly/admin/lock-workspace.spec.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    await aboutPage.getAboutEditIcon();
+    const workspaceEdit = new WorkspaceEditPage(page);
+    await workspaceEdit.waitForLoad();
+    await workspaceEdit.clickCancelButton();
+    // cleanup
+    await dataPage.deleteWorkspace();
+  });
+
+  // function to create cohort, dataset and notebook
+  async function createDatasetNotebook(page: Page, pyNotebookName: string): Promise<NotebookPreviewPage> {
```
Comment (aweng98) [inline e2e/tests/nightly/admin/lock-workspace.spec.ts:132]: This function is same as `workspace-admin-ui.spec` `createDatasetNotebook()`. I'm editing it in https://github.com/all-of-us/workbench/pull/6342. You can make this a common function in `test-utils.ts`.

**Review by aweng98 — CHANGES_REQUESTED** [review]

File: e2e/utils/test-utils.ts [hunk 11]
```diff
@@ -472,3 +473,17 @@ export async function findOrCreateCohort(page: Page, opts: { returnToDataPage?:
...
+
+  const datasetBuildPage = new DatasetBuildPage(page);
+  const exportModal = await datasetBuildPage.clickAnalyzeButton();
+
+  await exportModal.enterNotebookName(pyNotebookName);
+  await exportModal.pickLanguage(Language.Python);
+  await exportModal.clickExportButton();
+  const notebookPreviewPage = new NotebookPreviewPage(page);
+  return await notebookPreviewPage.waitForLoad();
+}
```
Comment (chavanr2019) [inline e2e/utils/test-utils.ts:489]: @aweng98 function moved from the lock-workspace and workspace-admin test file.

File: e2e/app/element/tab.ts [hunk 12]
```diff
@@ -20,8 +20,10 @@ export default class Tab extends BaseElement {
 
   async isSelected(): Promise<boolean> {
     const element = await this.asElementHandle();
-    const selected = await getPropValue<boolean>(element, 'aria-selected');
-    return selected ? selected : !!(await getStyleValue<string>(this.page, element, 'border-bottom'));
+    const selected = await getPropValue<string>(element, 'ariaSelected');
+    return selected !== undefined
+      ? selected === 'true'
+      : (await getStyleValue<string>(this.page, element, 'border-bottom')) !== null;
```
Comment (chavanr2019) [inline e2e/app/element/tab.ts:26]: @jmthibault79 tab file updated with the help of Alex, it gets the aria-selected status. Now the tab-status is verification works as expected.

**Review by chavanr2019 — COMMENTED** [review]
The review suggestion have been incorporated and with partial test steps I've been able to run the test successfully. But I would like to merge only when the createWorkspace issue is resolved. Please let me know if that works for both of you. Thanks!

File: e2e/app/page/workspace-about-page.ts [hunk 13]
```diff
@@ -73,4 +77,46 @@ export default class WorkspaceAboutPage extends WorkspaceBase {
...
+  // get the share button to verify if disabled for locked-workspace
+  getShareButton(): Button {
+    return Button.findByName(this.page, { containsText: 'Share' });
+  }
+
+  // get the REASON for lock-workspace
+  async getLockedWorkspaceReason(): Promise<string> {
+    const xpath = '//*[@data-test-id="lock-workspace-msg"]//child::div[2]/div[1]/b';
+    const element = BaseElement.asBaseElement(this.page, await this.page.waitForXPath(xpath, { visible: true }));
+    const textContent = element.getTextContent();
```
Comment (aweng98) [inline e2e/app/page/workspace-about-page.ts:95]: Missing a `await`.

**Review by aweng98 — APPROVED** [review]
LGTM!
